src/node/redeem.rs

```diff
@@ -273,6 +274,162 @@ impl<J: Jet> RedeemNode<J> {
         .unwrap()
     }
 
+    /// Prune the redeem program for the given transaction environment.
+    ///
+    /// Pruning works as follows:
+    /// 1) Run the redeem program on the Bit Machine.
+    /// 2) Mark all (un)used case branches using the IMR of the case node.
+    /// 3) Rebuild the program and omit unused branches.
+    ///
+    /// The pruning result depends on the witness data (which is already part of the redeem program)
+    /// and on the transaction environment. These two inputs determine which case branches are
+    /// used and which are not. Pruning must be done for each transaction environment separately,
+    /// starting from the same original, unpruned program. Pruning is a lossy process, so pruning
+    /// an already pruned program is not sound.
+    ///
+    /// Pruning fails if the original, unpruned program fails to run on the Bit Machine (step 1).
+    /// In this case, the witness data needs to be revised.
+    /// The other pruning steps (2 & 3) never fail.
+    pub fn prune(&self, env: &J::Environment) -> Result<Arc<RedeemNode<J>>, ExecutionError> {
+        struct Pruner<J> {
+            inference_context: types::Context,
+            executed_left: HashSet<Imr>,
+            executed_right: HashSet<Imr>,
+            phantom: PhantomData<J>,
+        }
+
+        impl<J: Jet> Converter<Redeem<J>, Witness<J>> for Pruner<J> {
+            type Error = std::convert::Infallible;
+
+            fn convert_witness(
+                &mut self,
+                data: &PostOrderIterItem<&RedeemNode<J>>,
+                witness: &Value,
+            ) -> Result<Option<Value>, Self::Error> {
+                let pruned = witness.prune(&data.node.arrow().target);
+                debug_assert!(
```

Comment:
In 7d93c829eed0ca9935c595214c5fa526564f4bf8:

I'd prefer these were full asserts. It's an extremely cheap check and if we mess up somehow so that it fails it will result in very mysterious failure modes.